File: cascadingFailuresTaskMigration_python/main/function.py
import math
from typing import Dict
import networkx as nx


class Function:
    def __init__(self, id_to_robots: Dict, id_to_groups: Dict):
        self.id_to_robots = id_to_robots
        self.id_to_groups = id_to_groups

    def calculate_over_load_is(self, robot) -> float:
        """Calculate Individual Survivability."""
        load = robot.load
        # Get group survivability score
        gs = self._calculate_gs(self.id_to_groups[robot.group_id])
        # Survivability function
        return max(gs * (1 - self._sig(load / 60)), 0.3)

    def _calculate_gs(self, group) -> float:
        """Calculate Group Survivability."""
        group_load = group.group_load
        # Use sigmoid-like function for monotonically non-increasing function between 0-1
        size = len(group.robot_id_in_group)
        return max(1 - self._sig(group_load / (size * 200)), 0.6)

    def _sig(self, x: float) -> float:
        """Sigmoid function variant."""
        return (math.exp(math.log(x + 1)) - math.exp(-math.log(x + 1))) / \
               (math.exp(math.log(x + 1)) + math.exp(-math.log(x + 1)))

    def calculate_contextual_load(self, leader, robot, arc_graph: nx.Graph,
                                  shortest_path_dict: Dict, a: float, b: float) -> float:
        """Calculate contextual load of a robot."""
        f = a * robot.load / robot.capacity - b * self.calculate_over_load_is(robot)

        # Get domain F from connected edges
        neighbors = list(arc_graph.neighbors(robot.robot_id))
        domain_f = 0.0
        cost_sum = 0.0

        for neighbor_id in neighbors:
            target_robot = self.id_to_robots[neighbor_id]

            if target_robot.group_id != robot.group_id or target_robot.robot_id == robot.robot_id:
                continue

            # Sum of communication costs with connected edges
            cost_sum += arc_graph[robot.robot_id][neighbor_id]['weight']
            domain_f += a * target_robot.load / target_robot.capacity - b * self.calculate_over_load_is(target_robot)

        size = len(neighbors) + 1
        domain_num = size + 1

        # Add cost for inter-layer task migration
        path_weight = shortest_path_dict.get((leader.robot_id, robot.robot_id), float('inf'))
        cost_sum += path_weight

        # Load function
        return f + 0.1 * (domain_f / domain_num + cost_sum / size)

File: cascadingFailuresTaskMigration/src/main/java/main/Function.java
package main;

import input.Group;
import input.Robot;
import org.jgrapht.alg.interfaces.ShortestPathAlgorithm;
import org.jgrapht.graph.DefaultUndirectedWeightedGraph;
import org.jgrapht.graph.DefaultWeightedEdge;

import java.util.HashMap;
import java.util.Set;

public class Function {
    HashMap<Integer, Robot> idToRobots;
    HashMap<Integer, Group> idToGroups;

    public Function(HashMap<Integer, Robot> idToRobots, HashMap<Integer, Group> idToGroups) {
        this.idToRobots = idToRobots;
        this.idToGroups = idToGroups;
    }

    public Double calculateOverLoadIS(Robot robot){
        //计算Individual Survivability
        double load = robot.getLoad();
        //获取组的生存评分
        double GS = calculateGS(idToGroups.get(robot.getGroupId()));
        //这里的存活率函数要好好设置一下
        return Math.max(GS*(1-sig(load/60)),0.3);

    }

    private Double calculateGS(Group group) {
        //计算Group Survivability
        double groupLoad = group.getGroupLoad();
        //用类似sigmod函数的变体做0-1之间的单调非增函数
        int size = group.getRobotIdInGroup().size();
        return Math.max(1-sig(groupLoad/(size*200)),0.6);
        //用来函数内进行调用
    }


    private Double sig(double x) {
        return (Math.exp(Math.log(x+1))-Math.exp(-Math.log(x+1)))/
                (Math.exp(Math.log(x+1))+Math.exp(-Math.log(x+1)));
    }

    public Double calculateContextualLoad(Robot leader, Robot robot,
                                          DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> arcGraph,
                                          ShortestPathAlgorithm<Integer, DefaultWeightedEdge> shortestPath,
                                          Double a , Double b) {
        double f = a*robot.getLoad()/robot.getCapacity()-b* calculateOverLoadIS(robot);
        //由相连边获取了domainF
        Set<DefaultWeightedEdge> defaultWeightedEdges = arcGraph.edgesOf(robot.getRobotId());
        double domianF = 0;
        double costSum = 0;
        for (DefaultWeightedEdge defaultWeightedEdge : defaultWeightedEdges) {
            Robot targetRobot = idToRobots.get(arcGraph.getEdgeTarget(defaultWeightedEdge));

            if (targetRobot.getGroupId()!=robot.getGroupId()||targetRobot.getRobotId()==robot.getRobotId()) {
                continue;
            }
            //与其相连边的沟通成本之和
            costSum+=arcGraph.getEdgeWeight(defaultWeightedEdge);
            domianF+=a*targetRobot.getLoad()/targetRobot.getCapacity()-b* calculateOverLoadIS(targetRobot);
        }
        int size = defaultWeightedEdges.size()+1;
        int domainNum = size +1;

        //加上进行层间任务迁移的cost
        double pathWeight = shortestPath.getPathWeight(leader.getRobotId(), robot.getRobotId());
        costSum+= pathWeight;
        //负载这个函数得想想
        return f+0.1*(domianF/domainNum+costSum/ size);
    }

}
